_source/_posts/2022-02-23-angular-async-config.md

```diff
@@ -0,0 +1,340 @@
+  ],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+**Summary**
+
+Customizing environment files is the Angular recommended method to inject values during build time.
+
+**When to use**
+
+* You have different configuration values based on build output
+
+**Best for**
+
+* Devmode - keep apps served locally from doing things only prod apps should do 
+* Multiple staging environment systems
+
+**Watch out for**
+
+* Configuration that involves private keys or tokens
+* Run your build for all the different environments when testing changes you make. You don't want to miss adding a property and potentially get a runtime error.
+
+## Loading configurations from external APIs
+
+Sometimes you need to load configuration at runtime. Doing so makes sense if you use release promotion style deployments - creating a build for a staging/pre-production environment and promoting the same build to production after verification. You don't want to create a new build, but what if your staging and production environments require different configurations? Loading configuration from an external API is handy in scenarios like these.
+
+I'll focus only on the Okta example for this configuration method to keep this post from getting too heavy.
+
+In this example, we'll look at `src/main.ts` where we bootstrap the Angular application. When you need configuration before the application loads, we can take advantage of `platformBrowserDynamic()` platform injector's `extraProviders` functionality to [provide platform providers](https://angular.io/guide/hierarchical-dependency-injection#platform-injector).
+
+Since we need to make the server call to get the configuration before we have a full Angular application context, we need to use Web APIs to call the API. Then we can configure the provider for Okta's `OKTA_CONFIG` injection token.
+
+For a configuration API call response that looks like this
+
+```json
+{
+  issuer: 'https://{yourOktaDomain}/oauth2/default',
+  clientId: '{yourClientId}', 
+  redirectUri: '{correctHostForTheEnvironment}/login/callback'
+}
+```
+
+The code in your `src/main.ts` changes to
+
+```ts
+import { enableProdMode } from '@angular/core';
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
+
+import { AppModule } from './app/app.module';
+import { environment } from './environments/environment';
+import { OKTA_CONFIG } from '@okta/okta-angular';
+import { OktaAuth } from '@okta/okta-auth-js';
+
+if (environment.production) {
+  enableProdMode();
+}
+
+fetch('http://{yourApiUri}/config').then(async res => {
+  const authConfig = await res.json();
+
+  platformBrowserDynamic([
+    { provide: OKTA_CONFIG, useValue: {oktaAuth: new OktaAuth(authConfig)}}
+  ]).bootstrapModule(AppModule)
+    .catch(err => console.error(err));
+});
+```
+
+Then your `AppModule` only needs to import `OktaAuthModule` since you already provided the `OKTA_CONFIG` injection token.
```

Comment:
```suggestion
Then your `AppModule` only needs to import `OktaAuthModule` since you've already provided the `OKTA_CONFIG` injection token.
```
past perfect tense - useful for series of steps that take place in a sequence in the past...